src/SignalR/server/Core/src/DefaultHubLifetimeManager.cs

```diff
@@ -271,14 +292,22 @@ public override Task SendUserAsync(string userId, string methodName, object?[] a
     public override Task OnConnectedAsync(HubConnectionContext connection)
     {
         _connections.Add(connection);
+        connection.Features.Set(new GroupTrackerFeature()); //add a group tracker to every concection
         return Task.CompletedTask;
     }
 
     /// <inheritdoc />
     public override Task OnDisconnectedAsync(HubConnectionContext connection)
     {
+        //now remove from tracked groups one by one
+        //this is faster than calling _groups.RemoveDisconnectedConnection
+        //because that method iteratas through ALL the groups
+        foreach (var grpName in connection.Features.GetRequiredFeature<GroupTrackerFeature>().Groups.ToArray()) //copy to array because groups can be modified in other methods, prevent "collection was modified"
```

Comment:
if you lock, you don't need `ToArray`

Reply:
Roger that, added a `lock` instead of `ToArray` (which iterates anyways)